docs/root/configuration/listeners/network_filters/generic_proxy_filter.rst

```diff
@@ -0,0 +1,128 @@
+and ensure they implement ``Request`` or ``Response``. It is much easier than implement a new network filter.
+
+
+Extendable matcher and route
+----------------------------
+
+:ref:`Generic matcher API <arch_overview_matching_api>` is used to construct the route table of the generic proxy. The developers could extend
+input and matcher to support new matching logic.
+
+By default, the generic proxy supports following input and matcher:
+
+* :ref:`host <envoy_v3_api_msg_extensions.filters.network.generic_proxy.matcher.v3.HostMatchInput>`: match the host of the request. The host should
+  represents set of instances that can be used for load balancing. It could be DNS name or VIP of the target service.
+* :ref:`path <envoy_v3_api_msg_extensions.filters.network.generic_proxy.matcher.v3.PathMatchInput>`: match the path of the request. The path should
+  represents RPC service name that used to represents set of method or functionality provided by target service.
+* :ref:`method <envoy_v3_api_msg_extensions.filters.network.generic_proxy.matcher.v3.MethodMatchInput>`: match the method of the request.
+* :ref:`property <envoy_v3_api_msg_extensions.filters.network.generic_proxy.matcher.v3.PropertyMatchInput>`: match the property of the request.
+  The property could be any property of the request that indexed by string key.
+* :ref:`request input <envoy_v3_api_msg_extensions.filters.network.generic_proxy.matcher.v3.RequestMatchInput>` and
+  :ref:`request matcher <envoy_v3_api_msg_extensions.filters.network.generic_proxy.matcher.v3.RequestMatcher>`: match the whole request by combining
+  host, path, method and properties in AND semantics. This is used to match multiple fields of the downstream request and avoid complex combinations of
+  host, path, method and properties in the generic matching tree.
+
+.. note::
+   Virtual methods are used to get the request properties in ``Request`` like ``host()``, ``path()``, ``method()``, etc. The developers need to
+   implement these virtual methods and determine what these fields are in their own protocol. It is possible that a protocol does not have some of
+   these fields. For example, the protocol does not have ``host`` field. In this case, the developers could return empty string in the virtual method
+   to indicate the field does not exist in the protocol. The only drawback is that the generic proxy could not match the request by this field.
+
+
+.. literalinclude:: _include/generic_proxy_filter.yaml
+    :language: yaml
+    :linenos:
+    :lineno-start: 21
+    :lines: 21-53
+
+Asynchronous codec API
+----------------------
+
+The generic proxy provides extension point to let the developers to implement specific codec for their own protocol. The codec API is designed to be
+asynchronous to avoid blocking the worker thread. And the asynchronous codec API make it is possible to accelerate the codec by offloading the parsing
```

Comment:
```suggestion
asynchronous to avoid blocking the worker thread. And the asynchronous codec API makes it possible to accelerate the codec by offloading the parsing
```